### PR TITLE
ports/stm32/rfcore.c: STM32WB BLE improvements

### PR DESCRIPTION
**Pull request by jimmo**
* Update to support newer WS versions (tested with WS=1.8). This fixes concurrent use of USB and BLE.
* Fix HCI parsing which lead to dropped HCI frames (now passes BLE multitests)
* Hook the RX IRQ for lower-latency BLE responses.
* Add a helper script for debugging the IPCC mechanism.

Also some other minor fixes and tidy-ups.

### CODE REVIEW

**Comment by jimmo** [comment]
@dmazzella FYI

**Comment by dmazzella** [comment]
@jimmo  tested and work for me.

**Comment by jimmo** [comment]
Rebased after merging #6405. Added RX IRQ handler.

File: ports/stm32/stm32_it.c [hunk 1]
```diff
@@ -828,7 +829,23 @@ void UART10_IRQHandler(void) {
 }
 #endif
 
-#endif
+#endif // USART for other families (non-FO/L0)
+
+// IPCC IRQ Handlers
+
+#if defined(STM32WB)
+void IPCC_C1_TX_IRQHandler(void) {
```
Comment (dpgeorge) [inline ports/stm32/stm32_it.c:837]: please move these IRQ handlers to rfcore.c

**Comment by dpgeorge** [comment]
I think the only CI failure is black which wants to reformat `rfcore.py`, so pls reformat it.

File: ports/stm32/make-stmconst.py [hunk 2]
```diff
@@ -160,7 +160,11 @@ def parse_file(filename):
             if m[0] == "}":
                 pass
             elif m[0] == "} TypeDef":
-                reg_defs[m[1].groupdict()["id"]] = regs
+                d = m[1].groupdict()
+                n = d["id"]
+                g = d["global"]
+                if n not in reg_defs or not g:
+                    reg_defs[m[1].groupdict()["id"]] = regs
```
Comment (dpgeorge) [inline ports/stm32/make-stmconst.py:167]: can this now be written as `reg_defs[n] = regs`?

File: ports/stm32/boards/USBDONGLE_WB55/mpconfigboard.mk [hunk 3]
```diff
@@ -1,9 +1,18 @@
...
+ifeq ($(USE_MBOOT),1)
+# When using Mboot all the text goes together after the bootloader
+LD_FILES = boards/stm32wb55xg.ld boards/common_bl.ld
+TEXT0_ADDR = 0x08004000
+else
+# When not using Mboot the text goes at the start of flash
+LD_FILES = boards/stm32wb55xg.ld boards/common_basic.ld
+TEXT0_ADDR = 0x08000000
+endif
+
```
Comment (dpgeorge) [inline ports/stm32/boards/USBDONGLE_WB55/mpconfigboard.mk:15]: this change really belongs in a separate (its own?) commit (not with "Update rfcore.c to match ST examples.")

File: ports/stm32/boards/stm32wb55xg.ld [hunk 4]
```diff
@@ -9,7 +9,8 @@ MEMORY
     FLASH_APP (rx)  : ORIGIN = 0x08004000, LENGTH = 496K /* sectors 4-127 */
     FLASH_FS (r)    : ORIGIN = 0x08080000, LENGTH = 256K /* sectors 128-191 */
     RAM (xrw)       : ORIGIN = 0x20000000, LENGTH = 192K /* SRAM1 */
-    RAM2A (xrw)     : ORIGIN = 0x20030020, LENGTH = 8K /* SRAM2A */
+    RAM2A (xrw)     : ORIGIN = 0x20030000, LENGTH = 10K  /* SRAM2A */
+    RAM2B (xrw)     : ORIGIN = 0x20038000, LENGTH = 10K  /* SRAM2B */
```
Comment (dpgeorge) [inline ports/stm32/boards/stm32wb55xg.ld:13]: there's a gap (of 22k) betwoon RAM2A and RAM2B, is that intentional?

File: ports/stm32/rfcore.c [hunk 5]
```diff
@@ -61,23 +81,92 @@ typedef struct _parse_hci_info_t {
...
+// [16:23] = Version minor
+// [24:31] = Version major
+
+// Memory Size
+// [0:7]   = Flash ( Number of 4k sector)
+// [8:15]  = Reserved ( Shall be set to 0 - may be used as flash extension )
+// [16:23] = SRAM2b ( Number of 1k sector)
+// [24:31] = SRAM2a ( Number of 1k sector)
+
+typedef struct __attribute__((packed)) {
```
Comment (dpgeorge) [inline ports/stm32/rfcore.c:97]: following convention elsewhere, these struct's should have, eg, `_mb_device_info_table_t` before the `{`

File: ports/stm32/rfcore.c [hunk 6]
```diff
@@ -61,23 +81,92 @@ typedef struct _parse_hci_info_t {
...
+    uint32_t safeboot_version;
+    uint32_t fus_version;
+    uint32_t fus_memorysize;
+    uint32_t fus_info;
+    uint32_t fw_version;
+    uint32_t fw_memorysize;
+    uint32_t fw_infostack;
+    uint32_t fw_reserved;
+} mb_device_info_table_t;
```
Comment (dpgeorge) [inline ports/stm32/rfcore.c:106]: what does `mb` stand for?

**Comment by dpgeorge** [comment]
The commit "Update rfcore.c to match ST examples." adds 2 commented-out lines that call LL_C2_EXTI_EnableEvent_32_63 and LL_EXTI_EnableRisingTrig_32_63, which are removed in commit "Minor refactoring.".  Please remove them from the first commit there.

File: ports/stm32/powerctrlboot.c [hunk 7]
```diff
@@ -154,12 +154,22 @@ void SystemClock_Config(void) {
...
+#define CLK48_SEMID (5)
+
 void SystemClock_Config(void) {
     // Enable the 32MHz external oscillator
     RCC->CR |= RCC_CR_HSEON;
     while (!(RCC->CR & RCC_CR_HSERDY)) {
     }
 
+    // Prevent CPU2 from disabling CLK48.
+    while (LL_HSEM_1StepLock(HSEM, CLK48_SEMID)) {}
```
Comment (dpgeorge) [inline ports/stm32/powerctrlboot.c:171]: please format like it's done above

File: ports/stm32/rfcore.c [hunk 8]
```diff
@@ -34,21 +35,40 @@
...
-#define IPCC_CH_BLE         (0x01) // BLE HCI command and response
-#define IPCC_CH_SYS         (0x02) // system HCI command and response
-#define IPCC_CH_MM          (0x08) // release buffer
-#define IPCC_CH_HCI_ACL     (0x20) // HCI ACL outgoing data
+#define IPCC_CH_BLE         (LL_IPCC_CHANNEL_1) // BLE HCI command and response
+#define IPCC_CH_SYS         (LL_IPCC_CHANNEL_2) // system HCI command and response
+#define IPCC_CH_MM          (LL_IPCC_CHANNEL_4) // release buffer
+#define IPCC_CH_HCI_ACL     (LL_IPCC_CHANNEL_6) // HCI ACL outgoing data
+
+#define OGF_CTLR_BASEBAND        (0x03)
```
Comment (dpgeorge) [inline ports/stm32/rfcore.c:50]: is this ctl-r (eg control register) or supposed to be ctrl?

Comment (dpgeorge) [reply ports/stm32/rfcore.c:50]: ahh, it looks like it's controller

File: ports/stm32/rfcore.c [hunk 9]
```diff
@@ -61,23 +81,92 @@ typedef struct _parse_hci_info_t {
...
+STATIC uint8_t ipcc_membuf_memmgr_sys_spare_evt_buf[272]; // mem2
+STATIC uint8_t ipcc_membuf_memmgr_evt_pool[6 * 272];  // mem2
+
+STATIC uint8_t ipcc_membuf_ble_cmd_buf[272]; // mem2
+STATIC uint8_t ipcc_membuf_ble_cs_buf[272]; // mem2
+STATIC tl_list_node_t ipcc_mem_ble_evt_queue; // mem1
+STATIC uint8_t ipcc_membuf_ble_hci_acl_data_buf[272]; // mem2
+
+// Set by the RX IRQ handler on incoming HCI payload.
+volatile STATIC bool had_ble_irq = false;
```
Comment (dpgeorge) [inline ports/stm32/rfcore.c:169]: please swap STATIC and volatile

File: ports/stm32/boards/stm32wb55xg.ld (hunk 4 above)
Comment (jimmo) [reply ports/stm32/boards/stm32wb55xg.ld:13]: It's a conservative estimate of the non-secure area of SRAM2A.

Comment (dpgeorge) [reply ports/stm32/boards/stm32wb55xg.ld:13]: Ok

**Comment by dpgeorge** [comment]
This was merged in b27edb8073409a3caee921dd4e2c1b182bd4c0fb through 632e3b7acc51b3b28f310ef5e23b0983c8e8b700 with some very minor clean ups.

**Pull request closed without merging**